autoprecompiles/src/bitwise_lookup_optimizer.rs

```diff
@@ -0,0 +1,139 @@
+            panic!();
+        };
+        if bus_int.mult != T::from(1).into() {
+            continue;
+        }
+
+        let AlgebraicExpression::Number(op) = op else {
+            continue;
+        };
+        // Collect all expressions to be byte-constrained and
+        // re-emit the constraints at the end in a (hopefully) more efficient way.
+        // The idea behind is that we can have duplicates or concrete numbers
+        // in those expressions.
+        if op == &0.into() {
+            // The bus interaction is equivalent to "x and y are bytes and z = 0".
+            to_byte_constrain.extend([x.clone(), y.clone()]);
+            to_remove.push(index);
+            if z != &T::from(0).into() {
+                new_constraints.push(z.clone().into());
+            }
+        } else if op == &1.into() {
+            // The bus interaction is equivalent to "x, y and z are bytes and z = x ^ y".
+            if x == &T::from(0).into() {
+                // y = z and both are bytes.
+                new_constraints.push((y.clone() - z.clone()).into());
+                to_byte_constrain.push(y.clone());
+                to_remove.push(index);
+            } else if y == &T::from(0).into() {
+                // x = z and both are bytes.
+                new_constraints.push((x.clone() - z.clone()).into());
+                to_byte_constrain.push(x.clone());
+                to_remove.push(index);
+            } else if z == &T::from(0).into() {
+                // x = y are both are bytes.
+                new_constraints.push((x.clone() - y.clone()).into());
+                to_byte_constrain.push(x.clone());
+                to_remove.push(index);
+            }
+        } else {
+            panic!("Expected bitwise bus interaction operation to be either 0 or 1.");
+        }
+    }
+    machine.bus_interactions = machine
+        .bus_interactions
+        .into_iter()
+        .enumerate()
+        .filter(|(i, _)| !to_remove.contains(i))
+        .map(|(_, bus_int)| bus_int)
+        .collect();
+    // After we have removed the bus interactions, we check which of the
+    // expressions we still need to byte-constrain are maybe already
+    // byte-constrained by other bus interactions.
+    let already_byte_constrained = all_byte_constrained_expressions(&machine)
+        .cloned()
+        .collect::<BTreeSet<_>>();
+    let mut to_byte_constrain = to_byte_constrain
+        .into_iter()
+        .filter(|expr| {
+            if let AlgebraicExpression::Number(n) = expr {
+                assert!(*n >= T::from(0) && *n < T::from(256));
+                // No need to byte-constrain numbers.
+                false
+            } else {
+                !already_byte_constrained.contains(expr)
+            }
+        })
+        .unique()
+        .collect_vec();
+    if to_byte_constrain.len() % 2 != 0 {
+        to_byte_constrain.push(T::from(0).into());
+    }
+    for chunk in to_byte_constrain.chunks(2) {
```

Comment:
is there like a `chunks_exact` thing that panics if not multiple of the argument?